### DISCUSSION TITLE
MCU Support for ESP32WROOM32d

### DISCUSSION THREAD

**Discussion by tobiasraak** [slint-ui/slint#5008] [Q&A]
Hello,

I really like the effort on the Esp-IDF integration and I was able to run the HelloWorld example from https://components.espressif.com/components/slint/slint on an ESP32Wroom32d by adjusting the `set-target` to `esp32` and the `add-dependency` to `esp_bsp_generic`. This works fine, since you can adjust all settings, pins for a display like the ili9341 in the menuconfig.
However I am quite stuck now. 
I want to try out the Printer-MCU-example but it somehow references the [MCU-Board-Support-package](https://github.com/slint-ui/slint/tree/master/examples/mcu-board-support) which only knows 4 specific boards that I do not want to use. And I get an exception right there since I set a different target in the first place.

Does anyone know if I can make use of the examples provided by slint to port them to a more generic esp32?
Furthermore, how do I proceed from here? All examples are made to use CMake or cargo which I do not use in esp-idf (technically indirect but I am not very familiar with it). The esp-idf doc is quite short. Is it possible to use rust code instead of cpp?

Thanks everyone in advance for any kind of advice!

**Comment by ogoffart** [comment]
I get from your comment that you want to use Rust.

Then you need to implement your own platform. Similar to what is done in this repository: https://github.com/slint-ui/slint-mcu-rust-template
(this one is for the rp2040, so all the code that is specific to that needs to be replaced by code that use the esp32-hal crate and the appropriate screen and touch driver)
You can use the code from https://github.com/slint-ui/slint/blob/master/examples/mcu-board-support/esp32_s3_box.rs for inspiration as it already is for a esp hardware.

**Reply by tobiasraak** [reply]
Thanks so much for the fast reply!
I am kind of open to using Rust or cpp, I really like the esp-idf integration since I am already familiar with it and I find it quite comfortable to configure the hardware parameters with it.
Taking a look at the the esp32-s3-box code I have to admit that I got no entry point on how to adapt this code to the esp32wroom.
If I, let's say don't want to get the printer sample running but instead move forward with my hello-world approach that is running, can I just proceed here: https://releases.slint.dev/1.5.1/docs/cpp/

Where is the point, where I need to make use of the mcu-board-support package or do I not need it at all (or only if I want to use rust)?

**Reply by ogoffart** [reply]
Using the esp-idf from C++ should be fairly easy.
You can try to follow the instructions in the readme in https://components.espressif.com/components/slint/slint
And see https://github.com/slint-ui/slint/tree/master/examples/printerdemo_mcu/esp-idf for an example.

**Reply by tobiraak** [reply]
Hey it's me again from my non-work account. I didn't had much time the past couple of months but I got a follow up question to that. Is it really mandatory to make use of **a** MCU board support package? What is the package for?

**Reply by tronical** [reply]
If you'd like to use Slint with Rust on this board, you're not required to use the mcu-board-support package. After all, it's just in our examples repo to keep them tidy and avoid code duplication, but the crate is not published on crates.io (yet?). You can look into the `esp32_s3_box.rs` for inspiration, it might make your life easier.

**Reply by tobiraak** [reply]
Thanks for the fast reply!
So you are referring to the mcu crates, right? 
Let's say I want to use rust with slint on an esp wroom32D, you said that it's not required to use the mcu-board-support package, however on the ["how to use" guide](https://github.com/slint-ui/slint/tree/master/examples/mcu-board-support) I need to init the mcu_board_support. Or is this just a generalized support for microcontrollers?
My end goal is to have the possibility to use slint to design my UI for an ili9341 with an esp32-wroom32d attached. Rust would be cool but is not mandatory, I don't mind writing cpp. What is the considered apporach to start?

**Reply by tronical** [reply]
> Thanks for the fast reply! So you are referring to the mcu crates, right? 

Yes.

> Let's say I want to use rust with slint on an esp wroom32D, you said that it's not required to use the mcu-board-support package, however on the ["how to use" guide](https://github.com/slint-ui/slint/tree/master/examples/mcu-board-support) I need to init the mcu_board_support. 

Yes, that guide applies to those who would like to use the crate.

> Or is this just a generalized support for microcontrollers?

The crate is an attempt to abstract away differences across different MCUs for use with our demos in Slint. It's not published to crates.io and there are not API guarantees.

> My end goal is to have the possibility to use slint to design my UI for an ili9341 with an esp32-wroom32d attached. Rust would be cool but is not mandatory, I don't mind writing cpp. What is the considered apporach to start?

I'd say that you have three options:

1) Attempt to use Rust on bare metal esp32. I don't know how much of this particular MCU and the ili9341 is supported by the bare-metal crates of esp-rs.
2) Attempt to use Rust on top of esp-idf. See also https://docs.esp-rs.org/book/overview/index.html . (1) is 2.1 in the Rust on ESP Book, and (2) is 2.2.
3) Use esp-idf with C/C++.

Let's look at how these options relate to Slint:

1) Here you need to implement your own `slint::platform::Platform`. You can be inspired by what's in the mcu-board-support crate, but you won't be able to use it out of the box on your MCU. This is the leanest option, but also the most complex one.
2) Here you also need to implement your own `slint::platform::Platform`, but the esp-idf APIs make it easier. See https://github.com/WilstonOreo/slint-esp-workshop/tree/main for an example.
3) Here you don't need to implement a Slint platform integration, we provide that as part of https://components.espressif.com/components/slint/slint/versions/1.7.1


All three options are valid in their own way and I'd say their choice depends on how lean of a result you'd like and how much you want to dig into Rust.

**Reply by tobiraak** [reply]
Awesome reply! So basically the third option is the one that I already did with the "Hello World" page of slint that I was able to run with the esp_bsp_generic support package. And the printer example won't run oc since they depend on rust. 

But I can start from here by writing the ui and the implementation in cpp and I would be good to go without the need of a special platform for my specific chip. I will test that! 

See here was my lack of understanding between the different components since I wanted to make use of any example application but didn't know that Rust was the part that made the use of a specific platform mandatory. Thanks for the clarification 👍
